File: src/lib/cp10-verify.ts
import { createServerFn } from "@tanstack/react-start";
import { getGraphAccessToken } from "./graph-auth";
import { classifyApprovalLanguage } from "./approval-classifier";
import { cp10Configs, resolveCp10Project, type Cp10ProjectConfig } from "./cp10-config";
import { connectRovoMcp } from "../mcp/rovo-mcp-client";
// Dynamically imported at each call site below, never at module top level —
// evidence-screenshot.ts pulls in playwright, which must never be eagerly
// evaluated by the browser (this module has no server-only build boundary,
// so Vite serves it to the client unbundled in dev).

export type Cp10PendingReason = "not_ready" | "wrong_sender" | "ambiguous";

export type Cp10Result = {
  project_id: string;
  jira_ticket: string;
  outcome_id: "CP10-C" | "CP10-Pending";
  pending_reason: Cp10PendingReason | null;
  approver_actor: string | null;
  confidence: number | null; // meaningful only for the Compliant case
  evidence_ref: string | null;
  evidence_screenshot: string | null; // approval-email screenshot filename — null only when no email was found
  timestamp: string;
};

export type GateState = "pending" | "passed" | "failed";
export type Cp10GateStatus = { jira: GateState; mail: GateState; llm: GateState };

export type Cp10VerifyResponse = {
  config: Cp10ProjectConfig | null;
  result: Cp10Result | null;
  reply: string;
  gates: Cp10GateStatus;
};

// In-memory "backend" for this POC, same pattern as CP38.
const resultStore = new Map<string, Cp10Result>();

function initialGates(): Cp10GateStatus {
  return { jira: "pending", mail: "pending", llm: "pending" };
}

function cp10ReplyText(
  cfg: Cp10ProjectConfig,
  result: Cp10Result,
  approverName: string | null,
  actualStatus?: string,
  wrongSenderEmail?: string | null,
): string {
  if (result.outcome_id === "CP10-C") {
    return `CP10 — CAB Approval: Compliant\n${cfg.jiraTicket} was approved by ${approverName ?? result.approver_actor}, the authorized CAB approver for ${cfg.projectLabel}. Confidence: ${result.confidence}%.`;
  }
  if (result.pending_reason === "not_ready") {
    return `CP10 — CAB Approval: Pending\n${cfg.jiraTicket} has not yet reached CAB approval stage${actualStatus ? ` (current status: ${actualStatus})` : ""}.`;
  }
  if (result.pending_reason === "wrong_sender") {
    if (wrongSenderEmail) {
      return `CP10 — CAB Approval: Pending\nAn email was received, but from ${wrongSenderEmail} — this is not the authorized CAB approver for ${cfg.projectLabel}.`;
    }
    return `CP10 — CAB Approval: Pending\nNo CAB approval email has been received yet for ${cfg.jiraTicket}.`;
  }
  // ambiguous
  return `CP10 — CAB Approval: Pending\nAn email was received from the authorized CAB approver, but the approval language was not explicit enough to confirm.`;
}

// Core flow, framework-agnostic. Mirrors CP38's structure deliberately: a
// cheap deterministic gate (Jira status) runs before an expensive one (mail
// search), sender identity is a deterministic string match, and the LLM
// only ever classifies language — never identity or eligibility — and only
// once both gates ahead of it have already passed.
export async function verifyCp10(cfg: Cp10ProjectConfig): Promise<Cp10VerifyResponse> {
  const timestamp = new Date().toISOString();
  const gates = initialGates();

  const client = await connectRovoMcp();
  try {
    // Step 3 — fetch the Jira ticket via Rovo MCP.
    const resourcesRes = await client.callTool({ name: "getAccessibleAtlassianResources", arguments: {} });
    const resourcesText = (resourcesRes.content as { text: string }[])[0]?.text ?? "{}";
    const cloudId = JSON.parse(resourcesText).data?.resources?.[0]?.cloudId as string | undefined;
    if (!cloudId) {
      return {
        config: cfg,
        result: null,
        gates,
        reply: "Copilot couldn't resolve an Atlassian site (no accessible resources returned).",
      };
    }

    const issueRes = await client.callTool({
      name: "getJiraIssue",
      arguments: { cloudId, issueIdOrKey: cfg.jiraTicket, view: "full" },
    });
    const issueText = (issueRes.content as { text: string }[])[0]?.text ?? "{}";
    if (issueRes.isError) {
      return { config: cfg, result: null, gates, reply: `Copilot couldn't fetch ${cfg.jiraTicket}: ${issueText}` };
    }
    const issue = JSON.parse(issueText).data;
    const statusName: string = issue?.fields?.status?.name ?? "Unknown";
    const reporterEmail: string | null = issue?.fields?.reporter?.emailAddress ?? null;
    const reporterName: string | null = issue?.fields?.reporter?.displayName ?? null;

    // Step 4 — Jira gate. Fails → stop here, mail is never even searched.
    if (statusName !== "Pending CAB Approval") {
      gates.jira = "failed";
      const result: Cp10Result = {
        project_id: cfg.projectId,
        jira_ticket: cfg.jiraTicket,
        outcome_id: "CP10-Pending",
        pending_reason: "not_ready",
        approver_actor: null,
        confidence: null,
        evidence_ref: null,
        evidence_screenshot: null,
        timestamp,
      };
      resultStore.set(cfg.projectId, result);
      return { config: cfg, result, gates, reply: cp10ReplyText(cfg, result, reporterName, statusName) };
    }
    gates.jira = "passed";

    if (!reporterEmail) {
      return {
        config: cfg,
        result: null,
        gates,
        reply: `Jira gate passed, but ${cfg.jiraTicket} has no reporter email to authorize against.`,
      };
    }

    // Step 5 — mail gate: search only runs because the Jira gate passed.
    const tokenResult = await getGraphAccessToken();
    if (!tokenResult.ok) {
      return {
        config: cfg,
        result: null,
        gates,
        reply: `Jira gate passed, but the mailbox couldn't be reached: ${tokenResult.error}`,
      };
    }
    const sharedMailbox = process.env.MS_SHARED_MAILBOX ?? process.env.MS_TEST_MAILBOX;
    if (!sharedMailbox) {
      return {
        config: cfg,
        result: null,
        gates,
        reply: "No shared mailbox configured (set MS_SHARED_MAILBOX or MS_TEST_MAILBOX in .env).",
      };
    }

    type GraphMessage = {
      id?: string;
      from?: { emailAddress?: { address?: string } };
      body?: { content?: string };
      subject?: string;
      receivedDateTime?: string;
    };
    let message: GraphMessage | undefined;
    try {
      const searchParams = new URLSearchParams();
      const orQuery = cfg.subjectMatches.map((s) => `subject:${s}`).join(" OR ");
      searchParams.set("$search", `"${orQuery}"`);
      searchParams.set("$top", "10");
      searchParams.set("$select", "id,from,body,subject,receivedDateTime");
      const searchRes = await fetch(
        `https://graph.microsoft.com/v1.0/users/${encodeURIComponent(sharedMailbox)}/messages?${searchParams.toString()}`,
        {
          headers: {
            Authorization: `Bearer ${tokenResult.accessToken}`,
            Prefer: 'outlook.body-content-type="text"',
            ConsistencyLevel: "eventual",
          },
        },
      );
      const searchData = (await searchRes.json()) as { value?: GraphMessage[]; error?: { message: string } };
      if (!searchRes.ok) {
        return {
          config: cfg,
          result: null,
          gates,
          reply: `Copilot couldn't search the mailbox: ${searchData.error?.message ?? `HTTP ${searchRes.status}`}`,
        };
      }
      const needles = cfg.subjectMatches.map((s) => s.toLowerCase());
      const candidates = (searchData.value ?? []).filter((m) => {
        const subj = (m.subject ?? "").toLowerCase();
        return needles.some((n) => subj.includes(n));
      });
      candidates.sort((a, b) => (b.receivedDateTime ?? "").localeCompare(a.receivedDateTime ?? ""));
      message = candidates[0];
    } catch (error) {
      return {
        config: cfg,
        result: null,
        gates,
        reply: `Mailbox search failed: ${error instanceof Error ? error.message : "unknown error"}`,
      };
    }

    // Step 6 — deterministic sender check. The PRD's pending_reason enum
    // (not_ready | wrong_sender | ambiguous) has no separate "no email"
    // value, so both "nothing found" and "found but wrong sender" map to
    // wrong_sender — the reply text below still distinguishes them for the
    // human reading it.
    if (!message) {
      gates.mail = "failed";
      const result: Cp10Result = {
        project_id: cfg.projectId,
        jira_ticket: cfg.jiraTicket,
        outcome_id: "CP10-Pending",
        pending_reason: "wrong_sender",
        approver_actor: null,
        confidence: null,
        evidence_ref: null,
        evidence_screenshot: null,
        timestamp,
      };
      resultStore.set(cfg.projectId, result);
      return { config: cfg, result, gates, reply: cp10ReplyText(cfg, result, reporterName, statusName, null) };
    }

    const sender = message.from?.emailAddress?.address ?? "";

    // Evidence of the email that was actually found — captured as soon as
    // any message exists, regardless of sender/language outcome, so a
    // wrong-sender or ambiguous Pending result still shows the auditor what
    // was received. Never used to establish eligibility or identity itself,
    // and a capture failure never affects the outcome — it just leaves
    // evidence_screenshot null.
    const screenshotFileName = message.id
      ? await (await import("./evidence-screenshot")).captureApprovalEmailScreenshot({
          checkpoint: "cp10",
          projectId: cfg.projectId,
          messageId: message.id,
          mailbox: sharedMailbox,
          accessToken: tokenResult.accessToken,
          subject: message.subject ?? "",
          from: sender,
          receivedDateTime: message.receivedDateTime ?? "",
        })
      : null;

    if (sender.toLowerCase() !== reporterEmail.toLowerCase()) {
      gates.mail = "failed";
      const result: Cp10Result = {
        project_id: cfg.projectId,
        jira_ticket: cfg.jiraTicket,
        outcome_id: "CP10-Pending",
        pending_reason: "wrong_sender",
        approver_actor: null,
        confidence: null,
        evidence_ref: message.id ?? null,
        evidence_screenshot: screenshotFileName,
        timestamp,
      };
      resultStore.set(cfg.projectId, result);
      return { config: cfg, result, gates, reply: cp10ReplyText(cfg, result, reporterName, statusName, sender) };
    }
    gates.mail = "passed";

    // Step 7 — LLM checks approval language, only now that sender matched.
    const classification = await classifyApprovalLanguage(message.body?.content ?? "", message.subject ?? "");
    if (!classification.ok) {
      return {
        config: cfg,
        result: null,
        gates,
        reply: `Sender verified, but the approval-language check failed: ${classification.error}`,
      };
    }
    gates.llm = classification.value.explicit ? "passed" : "failed";

    const result: Cp10Result = classification.value.explicit
      ? {
          project_id: cfg.projectId,
          jira_ticket: cfg.jiraTicket,
          outcome_id: "CP10-C",
          pending_reason: null,
          approver_actor: reporterEmail,
          confidence: classification.value.confidence,
          evidence_ref: message.id ?? null,
          evidence_screenshot: screenshotFileName,
          timestamp,
        }
      : {
          project_id: cfg.projectId,
          jira_ticket: cfg.jiraTicket,
          outcome_id: "CP10-Pending",
          pending_reason: "ambiguous",
          approver_actor: reporterEmail,
          confidence: classification.value.confidence,
          evidence_ref: message.id ?? null,
          evidence_screenshot: screenshotFileName,
          timestamp,
        };

    resultStore.set(cfg.projectId, result);
    return { config: cfg, result, gates, reply: cp10ReplyText(cfg, result, reporterName, statusName) };
  } finally {
    await client.close();
  }
}

export const runCp10Verification = createServerFn({ method: "POST" })
  .validator((data: { question: string }) => data)
  .handler(async ({ data }): Promise<Cp10VerifyResponse> => {
    const cfg = resolveCp10Project(data.question);
    if (!cfg) {
      return {
        config: null,
        result: null,
        gates: initialGates(),
        reply: `I couldn't tell which project you mean. Known projects: ${cp10Configs
          .map((c) => c.projectLabel)
          .join(", ")}.`,
      };
    }
    return verifyCp10(cfg);
  });

export const getCp10LatestResult = createServerFn({ method: "GET" })
  .validator((data: { projectId: string }) => data)
  .handler(async ({ data }): Promise<Cp10Result | null> => resultStore.get(data.projectId) ?? null);

export const getCp10EvidenceScreenshot = createServerFn({ method: "GET" })
  .validator((data: { fileName: string }) => data)
  .handler(async ({ data }): Promise<{ dataUrl: string } | null> => {
    const { readEvidenceScreenshot } = await import("./evidence-screenshot");
    const buffer = await readEvidenceScreenshot(data.fileName);
    if (!buffer) return null;
    return { dataUrl: `data:image/png;base64,${buffer.toString("base64")}` };
  });

File: backend/app/services/cp10_verify.py
# Core flow, framework-agnostic. Mirrors CP38's structure deliberately: a
# cheap deterministic gate (Jira status) runs before an expensive one (mail
# search), sender identity is a deterministic string match, and the LLM
# only ever classifies language — never identity or eligibility — and only
# once both gates ahead of it have already passed.
#
# Jira access is direct REST + Basic auth (app/clients/jira_client.py), same
# as CR-compliance — the TS version's Atlassian Rovo MCP/OAuth flow was
# dropped in the Python port; no interactive login step required here.
import httpx

from app.clients.approval_classifier import classify_approval_language
from app.clients.evidence_screenshot import capture_approval_email_screenshot
from app.clients.graph_auth import get_graph_access_token
from app.clients.jira_client import get_jira_issue
from app.models import Cp10GateStatus, Cp10ProjectConfig, Cp10Result, Cp10VerifyResponse
from app.settings import settings
from app.utils import now_iso


def _initial_gates() -> Cp10GateStatus:
    return Cp10GateStatus(jira="pending", mail="pending", llm="pending")


def cp10_reply_text(
    cfg: Cp10ProjectConfig,
    result: Cp10Result,
    approver_name: str | None,
    actual_status: str | None = None,
    wrong_sender_email: str | None = None,
) -> str:
    if result.outcome_id == "CP10-C":
        approver = approver_name or result.approver_actor
        return (
            f"CP10 — CAB Approval: Compliant\n"
            f"{cfg.jira_ticket} was approved by {approver}, the authorized CAB approver for "
            f"{cfg.project_label}. Confidence: {result.confidence}%."
        )
    if result.pending_reason == "not_ready":
        status_note = f" (current status: {actual_status})" if actual_status else ""
        return f"CP10 — CAB Approval: Pending\n{cfg.jira_ticket} has not yet reached CAB approval stage{status_note}."
    if result.pending_reason == "wrong_sender":
        if wrong_sender_email:
            return (
                f"CP10 — CAB Approval: Pending\n"
                f"An email was received, but from {wrong_sender_email} — this is not the authorized "
                f"CAB approver for {cfg.project_label}."
            )
        return f"CP10 — CAB Approval: Pending\nNo CAB approval email has been received yet for {cfg.jira_ticket}."
    # ambiguous
    return (
        f"CP10 — CAB Approval: Pending\n"
        f"An email was received from the authorized CAB approver, but the approval language was not "
        f"explicit enough to confirm."
    )


async def verify_cp10(cfg: Cp10ProjectConfig) -> Cp10VerifyResponse:
    timestamp = now_iso()
    gates = _initial_gates()

    # Step 3 — fetch the Jira ticket via direct REST.
    try:
        issue = await get_jira_issue(cfg.jira_ticket)
    except RuntimeError as error:
        return Cp10VerifyResponse(config=cfg, result=None, gates=gates, reply=f"Copilot couldn't reach Jira: {error}")
    if not issue:
        return Cp10VerifyResponse(
            config=cfg, result=None, gates=gates, reply=f"Copilot couldn't find {cfg.jira_ticket} in Jira."
        )

    fields = issue.get("fields") or {}
    status_name: str = (fields.get("status") or {}).get("name") or "Unknown"
    reporter = fields.get("reporter") or {}
    reporter_email: str | None = reporter.get("emailAddress")
    reporter_name: str | None = reporter.get("displayName")

    # Step 4 — Jira gate. Fails → stop here, mail is never even searched.
    if status_name != "Pending CAB Approval":
        gates.jira = "failed"
        result = Cp10Result(
            project_id=cfg.project_id,
            jira_ticket=cfg.jira_ticket,
            outcome_id="CP10-Pending",
            pending_reason="not_ready",
            approver_actor=None,
            confidence=None,
            evidence_ref=None,
            evidence_screenshot=None,
            timestamp=timestamp,
        )
        return Cp10VerifyResponse(
            config=cfg, result=result, gates=gates, reply=cp10_reply_text(cfg, result, reporter_name, status_name)
        )
    gates.jira = "passed"

    if not reporter_email:
        return Cp10VerifyResponse(
            config=cfg,
            result=None,
            gates=gates,
            reply=f"Jira gate passed, but {cfg.jira_ticket} has no reporter email to authorize against.",
        )

    # Step 5 — mail gate: search only runs because the Jira gate passed.
    token_result = await get_graph_access_token()
    if not token_result.ok:
        return Cp10VerifyResponse(
            config=cfg,
            result=None,
            gates=gates,
            reply=f"Jira gate passed, but the mailbox couldn't be reached: {token_result.error}",
        )
    shared_mailbox = settings.ms_shared_mailbox or settings.ms_test_mailbox
    if not shared_mailbox:
        return Cp10VerifyResponse(
            config=cfg,
            result=None,
            gates=gates,
            reply="No shared mailbox configured (set MS_SHARED_MAILBOX or MS_TEST_MAILBOX in .env).",
        )

    message = None
    try:
        or_query = " OR ".join(f"subject:{s}" for s in cfg.subject_matches)
        params = {"$search": f'"{or_query}"', "$top": "10", "$select": "id,from,body,subject,receivedDateTime"}
        async with httpx.AsyncClient() as client:
            search_res = await client.get(
                f"https://graph.microsoft.com/v1.0/users/{shared_mailbox}/messages",
                params=params,
                headers={
                    "Authorization": f"Bearer {token_result.access_token}",
                    "Prefer": 'outlook.body-content-type="text"',
                    "ConsistencyLevel": "eventual",
                },
            )
        search_data = search_res.json()
        if search_res.status_code >= 400:
            error = (search_data.get("error") or {}).get("message") or f"HTTP {search_res.status_code}"
            return Cp10VerifyResponse(
                config=cfg, result=None, gates=gates, reply=f"Copilot couldn't search the mailbox: {error}"
            )
        needles = [s.lower() for s in cfg.subject_matches]
        candidates = [
            m for m in (search_data.get("value") or []) if any(n in (m.get("subject") or "").lower() for n in needles)
        ]
        candidates.sort(key=lambda m: m.get("receivedDateTime") or "", reverse=True)
        message = candidates[0] if candidates else None
    except httpx.HTTPError as error:
        return Cp10VerifyResponse(config=cfg, result=None, gates=gates, reply=f"Mailbox search failed: {error}")

    # Step 6 — deterministic sender check. The PRD's pending_reason enum
    # (not_ready | wrong_sender | ambiguous) has no separate "no email"
    # value, so both "nothing found" and "found but wrong sender" map to
    # wrong_sender.
    if not message:
        gates.mail = "failed"
        result = Cp10Result(
            project_id=cfg.project_id,
            jira_ticket=cfg.jira_ticket,
            outcome_id="CP10-Pending",
            pending_reason="wrong_sender",
            approver_actor=None,
            confidence=None,
            evidence_ref=None,
            evidence_screenshot=None,
            timestamp=timestamp,
        )
        return Cp10VerifyResponse(
            config=cfg,
            result=result,
            gates=gates,
            reply=cp10_reply_text(cfg, result, reporter_name, status_name, None),
        )

    sender = (message.get("from") or {}).get("emailAddress", {}).get("address") or ""
    message_id = message.get("id")

    # Evidence of the email that was actually found — captured as soon as
    # any message exists, regardless of sender/language outcome.
    screenshot_file_name = None
    if message_id:
        screenshot_file_name = await capture_approval_email_screenshot(
            checkpoint="cp10",
            project_id=cfg.project_id,
            message_id=message_id,
            mailbox=shared_mailbox,
            access_token=token_result.access_token,
            subject=message.get("subject") or "",
            from_=sender,
            received_date_time=message.get("receivedDateTime") or "",
        )

    if sender.lower() != reporter_email.lower():
        gates.mail = "failed"
        result = Cp10Result(
            project_id=cfg.project_id,
            jira_ticket=cfg.jira_ticket,
            outcome_id="CP10-Pending",
            pending_reason="wrong_sender",
            approver_actor=None,
            confidence=None,
            evidence_ref=message_id,
            evidence_screenshot=screenshot_file_name,
            timestamp=timestamp,
        )
        return Cp10VerifyResponse(
            config=cfg,
            result=result,
            gates=gates,
            reply=cp10_reply_text(cfg, result, reporter_name, status_name, sender),
        )
    gates.mail = "passed"

    # Step 7 — LLM checks approval language, only now that sender matched.
    classification = await classify_approval_language(
        (message.get("body") or {}).get("content") or "", message.get("subject") or ""
    )
    if not classification.ok:
        return Cp10VerifyResponse(
            config=cfg,
            result=None,
            gates=gates,
            reply=f"Sender verified, but the approval-language check failed: {classification.error}",
        )
    gates.llm = "passed" if classification.value.explicit else "failed"

    if classification.value.explicit:
        result = Cp10Result(
            project_id=cfg.project_id,
            jira_ticket=cfg.jira_ticket,
            outcome_id="CP10-C",
            pending_reason=None,
            approver_actor=reporter_email,
            confidence=classification.value.confidence,
            evidence_ref=message_id,
            evidence_screenshot=screenshot_file_name,
            timestamp=timestamp,
        )
    else:
        result = Cp10Result(
            project_id=cfg.project_id,
            jira_ticket=cfg.jira_ticket,
            outcome_id="CP10-Pending",
            pending_reason="ambiguous",
            approver_actor=reporter_email,
            confidence=classification.value.confidence,
            evidence_ref=message_id,
            evidence_screenshot=screenshot_file_name,
            timestamp=timestamp,
        )

    return Cp10VerifyResponse(
        config=cfg, result=result, gates=gates, reply=cp10_reply_text(cfg, result, reporter_name, status_name)
    )
